File: Sphere Online Judge/A needle in the Haystack/A_Needle_in_the_Haystack.py
#  SPOJ - NHAY - A Needle in the Haystack
#  https://www.spoj.com/problems/NHAY/

def maximal_boudaries(w):
    k = 0
    n = len(w)
    f = [0]*n
    for i in range(1, n):
        while k > 0 and w[k] != w[i]:
            k = f[k-1]
        if w[k] == w[i]:
            k += 1
        f[i] = k
        
    return f

def solve(params):
    l, t, s = params
    f = maximal_boudaries(t+"\x00"+s)
    ans = []
    for i, n in enumerate(f):
        if n == len(t):
            ans.append(i-2*len(t))
    print(*ans, sep="\n")

def main():
    params = []
    for line in sys.stdin:
        params.append(line.rstrip())
        if len(params) == 3:
            solve(params)
            params = []

























import os, sys; from io import BytesIO, IOBase; 
class FastIO(IOBase):
    newlines = 0
    def __init__(self, file): self._fd = file.fileno(); self.buffer = BytesIO(); self.writable = "x" in file.mode or "r" not in file.mode; self.write = self.buffer.write if self.writable else None; self.BUFSIZE = 8192
    def readline(self):
        while self.newlines == 0: b = os.read(self._fd, max(os.fstat(self._fd).st_size, self.BUFSIZE)); self.newlines = b.count(b"\n") + (not b); ptr = self.buffer.tell(); self.buffer.seek(0, 2); self.buffer.write(b); self.buffer.seek(ptr)
        self.newlines -= 1; return self.buffer.readline()
class IOWrapper(IOBase):
    def __init__(self, file): self.buffer = FastIO(file); self.flush = self.buffer.flush; self.writable = self.buffer.writable; self.write = lambda s: self.buffer.write(s.encode("ascii")); self.read = lambda: self.buffer.read().decode("ascii"); self.readline = lambda: self.buffer.readline().decode("ascii")
sys.stdin = IOWrapper(sys.stdin)
def input(): return sys.stdin.readline().rstrip()

from math import *
from random import *
from collections import *
from bisect import *
from heapq import *

___ = getrandbits(31)
class myint(int):
    def __hash__(self):
        return self^___
int = myint

sys.setrecursionlimit(6000)

if __name__ == "__main__":
    main()

File: Sphere Online Judge/A needle in the Haystack/A_Needle_in_the_Haystack.cpp
//  SPOJ - NHAY - A Needle in the Haystack
//  https://www.spoj.com/problems/NHAY/

#include <bits/stdc++.h>
using namespace std;

vector<int> findOcurrences(int len_needle, string &haystack, string &needle) {
    string KMPString = needle + '\x00' + haystack;

    int k = 0, N = KMPString.size();

    vector<int> LPS(N), ocurrences;

    for(int i = 1; i < N; ++i) {
        while (k > 0 and KMPString[i] != KMPString[k]) {
            k = LPS[k -1];
        }

        k += KMPString[i] == KMPString[k];
        
        LPS[i] = k;

        if (LPS[i] == len_needle) {
            ocurrences.push_back(i - len_needle*2);
        }
    }

    return ocurrences;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int len_needle;
    string haystack, needle;

    while (cin >> len_needle) {
        cin >> needle;
        cin >> haystack;

        vector<int> ocurrences = findOcurrences(len_needle, haystack, needle);

        for (int i : ocurrences) {
            cout << i << "\n";
        }

        cout << "\n";
    }
}